flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/metrics/lifecycle/LifecycleMetrics.java

```diff
@@ -0,0 +1,217 @@
+                    histos.put(
+                            metricName,
+                            List.of(
+                                    t.f0,
+                                    t.f1.computeIfAbsent(
+                                            cr.getMetadata().getNamespace(),
+                                            ns ->
+                                                    createTransitionHistogram(
+                                                            metricName,
+                                                            operatorMetricGroup
+                                                                    .createResourceNamespaceGroup(
+                                                                            configManager
+                                                                                    .getDefaultConfig(),
+                                                                            ns)))));
+                });
+        return histos;
+    }
+
+    private Histogram createTransitionHistogram(String metricName, MetricGroup group) {
+        return metricGroupFunction
+                .apply(group)
+                .addGroup("Transition")
+                .addGroup(metricName)
+                .histogram(
+                        "TimeSeconds",
+                        new DescriptiveStatisticsHistogram(
+                                configManager
+                                        .getOperatorConfiguration()
+                                        .getMetricsHistogramSampleSize()));
+    }
+
+    private static List<Transition> getTrackedTransitions() {
+        return List.of(
+                new Transition(CREATED, DEPLOYED, false, TRANSITION_FIRST_DEPLOYMENT),
+                new Transition(SUSPENDED, STABLE, true, TRANSITION_RESUME),
+                new Transition(STABLE, STABLE, true, TRANSITION_UPGRADE),
+                new Transition(DEPLOYED, UPGRADING, true, TRANSITION_SUSPEND),
+                new Transition(STABLE, UPGRADING, true, TRANSITION_SUSPEND),
+                new Transition(ROLLED_BACK, UPGRADING, true, TRANSITION_SUSPEND),
+                new Transition(DEPLOYED, SUSPENDED, true, TRANSITION_SUSPEND),
+                new Transition(STABLE, SUSPENDED, true, TRANSITION_SUSPEND),
+                new Transition(ROLLED_BACK, SUSPENDED, true, TRANSITION_SUSPEND),
+                new Transition(DEPLOYED, STABLE, false, TRANSITION_STABILIZATION),
+                new Transition(DEPLOYED, ROLLED_BACK, false, TRANSITION_ROLLBACK),
+                new Transition(UPGRADING, DEPLOYED, true, TRANSITION_SUBMISSION),
+                new Transition(ROLLING_BACK, ROLLED_BACK, true, TRANSITION_SUBMISSION));
+    }
+
+    /**
+     * Pojo for encapsulating state transitions and whether we should measure time from the
+     * beginning of from or since the last update.
+     */
+    @ToString
+    @RequiredArgsConstructor
+    protected static class Transition {
```

Comment:
Have you considered defining `Transition` as enum?

Reply:
There are 12 instances of Transition currently with not very well defined names. Also they are treated in the same way, so I feel making them an enum is not a real good fit here.

Do you have a concrete suggestion?

Reply:
We always measure the latency between two states, so it could be STABLE_STABLE, SUPENDED_STABLE, etc. The last param shows which metric they belong to, not really which transition.

Reply:
but I might be missing something

Reply:
Enums only make sense in cases where the different values have different meanings. (and that actually means you have logic built around them)

Transition objects are handled uniformly, you can add / remove new ones later it won't change anything.

I think enum is not a good fit here.